### PR TITLE
Create colors.js for $mdColor service 

### PR DESCRIPTION
**Pull request by SanderElias**
mdColors service for MD

Reformat of PR https://github.com/angular/material/pull/1566, for issue #1269

[Review on Reviewable](https://reviewable.io/reviews/angular/material/1599)

### CODE REVIEW

**Comment by marcysutton** [comment]
@SanderElias we're going to start looking at this PR again now that ngConf is done, sorry for the delay. If you're able, can you also contribute some test coverage for this? 

@rschmukler this would benefit from your review as well.

**Comment by marcysutton** [comment]
@SanderElias are you planning to write any tests? @rschmukler you had some concerns about copying the Palettes object but weren't sure if it was required due to timing of theming builds. Can you take a closer look since you're more familiar with theming?

**Comment by SanderElias** [comment]
@marcy, my current workload does not really allow me enough time to dig into tests deep enough. Frankly I lack the time/experience to build the scaffolding for the tests. If the scaffolding is in place, I can write some tests. 
I agree with the copying, I will take another look at that. It was bothering me when I wrote it, but I didn't see an way around it back then. I'll dig in later today, and will come up with the reasoning behind it

**Comment by marcysutton** [comment]
Fair enough @SanderElias, thanks so much for all your work on this already!

**Comment by TalPasi** [comment]
Will the $mdColors will be aware if the theme was set as a Dark theme?

**Comment by rschmukler** [comment]
@SanderElias have you had any time to look at the copying? We'd love to get this into `0.9.0`. Once you've responded with your considerations, I can write some scaffolding for the tests and if you have time to flush them out, then great, otherwise I'll get to them when I can, but it might be after 0.9.0.

**Comment by SanderElias** [comment]
@rschmukler I did look into the copying just now. This can be addressed once #1584 is taken care of, depending of the outcome of that, we need a full copy, or I can adapt the code a bit to use the constant. 
The memory profile for the copy is very low, especially if you look at it in perspective of the entire framework.
The only other solution is traversing the service every time there is an color-change. I think this is more costly, especially on mobile then it is to retain a copy.

**Comment by marcysutton** [comment]
@SanderElias @rschmukler we really want to merge this, can you both chime in on its readiness?

**Comment by rschmukler** [comment]
As per @SanderElias comments, I'm good with the implementation. That said, I haven't had time to write tests for it yet.

**Comment by asans** [comment]
I'm wondering.  Is it possible to also include accent colors.  Currently, we have fg and bg colors based off the primary color palette.  However, it would also make sense to be able to use the accent colors, ie:

<p class="md-accent-1">Text</p>

Or perhaps some other more conforming style name.  Considering that only select elements support md-accent with md-hue-1/2/3, this may be a good addition.

**Comment by nathanmarks** [comment]
Is this still waiting for something to be completed? Or is it just waiting to be merged.

**Comment by SanderElias** [comment]
@nathanmarks, It just needs some tests. otherwise it's fine. I'm still lacking the time/experience to set up the test-scaffolding for this, and the same seems to go for @rschmukler. (well, in his case just the time part!) Once the tests are in place, it can be merged, and extended upon.

**Comment by mprentic-harris** [comment]
I've been trying to find the time to add some tests for this, but when I last tried, it wasn't clear that this still worked or how it should be used. Is there current documentation for this anywhere? Does it work with master?

**Comment by SanderElias** [comment]
@mprentic-harris, I just checked, it should work on master, but I had some conversation with Ryan, and I need to add support for multiple color-schemes, (primary/sceondary/etc) and some init functions. Will try to fit this in this week.

File: src/core/services/theming/colors.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            return ;
+
+            function addHue(hueName) {
+                colors.push({
+                    color: hueName,
+                    value: palette[hueColors[hueName]]
+                });
+            }
+
+            function copyColors(colorName) {
```
Comment (danwkennedy) [inline src/core/services/theming/colors.js]: Shouldn't "colorName" be "paletteName"? This method gets called on a hash of palettes.

**Comment by mackelito** [comment]
Is there any update on this PR?

**Comment by nathanmarks** [comment]
@ThomasBurleson What work is needed to get this done? I'm not 100% on the status after reading through and can possibly contribute to it.

**Comment by whittssg** [comment]
Did this make it in yet? Thanks

**Comment by ThomasBurleson** [comment]
@nathanmarks - This is an excellent idea from @SanderElias. Let me think about the overall API that we need. That will identify what extra work is needed with this PR.

**Comment by gkalpak** [comment]
Has this been superseded by #7791 ? (I'm not sure if it's exactly the same funtionality.)

**Comment by EladBezalel** [comment]
@gkalpak from the description it seems like it, 
Funny thing is that it's the first time i encounter this code.

@SanderElias thanks but closing as for #7791

**Comment by ThomasBurleson** [comment]
@EladBezalel - this has extra features that you may want to merge with `$mdColors`.

**Comment by ThomasBurleson** [comment]
Outdated or not core focus for Material 1.

**Pull request closed without merging**